### PR TITLE
Transcendental functions check model

### PR DESCRIPTION
**Pull request by ajreynol**
This pull request:
- Adds support for a check model using the error bounds from the Taylor approximation of transcendental functions, thus enabling us to answer "SAT" for transcendental functions with irrational values, without providing deltas on model values.
- Adds a scheme for incrementally increasing the precision of the Taylor series.
Both schemes follow the approaches described in "Satisfiability Modulo Transcendental Functions via Incremental Linearization".

### CODE REVIEW

File: src/options/arith_options [hunk 1]
```diff
@@ -195,4 +195,10 @@ option nlExtPurify --nl-ext-purify bool :default false
 option nlExtSplitZero --nl-ext-split-zero bool :default false
  intial splits on zero for all variables
 
+option nlExtTfTaylorDegree --nl-ext-tf-taylor-deg=N int16_t :default 4 :read-write
+ initial degree of polynomials for Taylor approximation
+
```
Comment (HanielB) [inline src/options/arith_options]: trailing space

File: src/options/arith_options [hunk 2]
```diff
@@ -195,4 +195,10 @@ option nlExtPurify --nl-ext-purify bool :default false
 option nlExtSplitZero --nl-ext-split-zero bool :default false
  intial splits on zero for all variables
 
+option nlExtTfTaylorDegree --nl-ext-tf-taylor-deg=N int16_t :default 4 :read-write
+ initial degree of polynomials for Taylor approximation
+ 
+option nlExtTfIncPrecision --nl-ext-tf-inc-prec bool :default true
+ whether to increment the precision for transcendental function constraints
+
```
Comment (HanielB) [inline src/options/arith_options]: trailing space

File: src/theory/arith/nonlinear_extension.h [hunk 3]
```diff
@@ -228,10 +228,48 @@ class NonlinearExtension {
...
+   * true in the current model. Notice that we typically cannot compute concrete
+   * values for assertions involving transcendental functions. Any assertion
+   * whose model value cannot be computed is included in the return value of
+   * this function.
+   */
+  std::vector<Node> checkModel(const std::vector<Node>& assertions);
+
+  /** check model for transcendental functions
+   *
+   * Check the model using error bounds on the Taylor approximation.
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: Check -> Checks

File: src/theory/arith/nonlinear_extension.h [hunk 4]
```diff
@@ -228,10 +228,48 @@ class NonlinearExtension {
...
+   *
+   * If this function returns true, then all assertions in the input argument
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: clarify to what "input argument" refers

File: src/theory/arith/nonlinear_extension.h [hunk 5]
```diff
@@ -228,10 +228,48 @@ class NonlinearExtension {
...
+   * their error bounds (as stored in d_tf_check_model_bounds).
+   *
+   * For details, see Section 3 of Cimatti et al CADE 2017 under the heading
+   * "Detecting Satisfiable Formulas".
+   */
+  bool checkModelTf(const std::vector<Node>& assertions);
+
+  /** simple check model for transcendental functions for literal
+   *
+   * This method returns true if literal is true for all interpretations of
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: interpretations of --> interpretations

File: src/theory/arith/nonlinear_extension.cpp [hunk 6]
```diff
@@ -1128,6 +1130,132 @@ std::set<Node> NonlinearExtension::getFalseInModel(
...
+    }
+  }
+
+  if (check_assertions.empty())
+  {
+    return true;
+  }
+  else
+  {
+    // TODO
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: Clarify what this TODO is, ideally linking with an issue.

**Comment by ajreynol** [comment]
I've addressed the review in my latest commit.

**Pull request merged**